pkg/storage/util.go

```diff
@@ -106,23 +104,19 @@ func NamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
-		return "", fmt.Errorf("invalid name: %v", msgs)
-	}
-	return prefix + "/" + meta.GetNamespace() + "/" + name, nil
+	// We assume that NamespaceKeyFunc is used for read requests (List and Watch)
+	// and thus the object is correct and we don't perform any validation.
+	return prefix + "/" + meta.GetNamespace() + "/" + meta.GetName(), nil
 }
 
 func NoNamespaceKeyFunc(prefix string, obj runtime.Object) (string, error) {
 	meta, err := meta.Accessor(obj)
 	if err != nil {
 		return "", err
 	}
-	name := meta.GetName()
-	if msgs := validation.IsValidPathSegmentName(name); len(msgs) != 0 {
```

Comment:
> @liggitt - after my changes to the cacher from ~1-2 months ago (that introduced trigger function), we really compute it only for a small number of watcher. So it's not a matter of huge number of them.

it looks like that helps watches that filter on label/field selectors, but not ones that filter on key prefix, so most watches (that don't specify a labelSelector or fieldSelector) would still have to filter all watch events from all namespaces, and they each currently invoke keyFunc() on each watch event.

to verify that, I did this:
1. started a cluster with logging in this function
2. opened 10 watches on configmaps in ns foo, and 10 across all namespaces:
```
kubectl create ns foo
kubectl get configmaps --namespace=foo --watch (10x)
kubectl get configmaps --all-namespaces --watch (10x)
```
3. created/deleted configmaps in ns "default"
```
kubectl create configmap mymap --namespace=default
kubectl delete configmap mymap --namespace=default
```

This function fired 20 times, once for each watcher, even though 10 were watching a completely different namespace. If we computed the object key once when the cache received the watch event, before fanning out to the watchers, we'd avoid repeated calls to this function.

> Actually, I'm slightly lost in this discussion. I completely agree that we really need such validation at the level of writing to storage. But why do we need to check it on read?

I'm not opposed to optimizing this, but I think the repeated calls are the main problem.

I actually think we should work to make the cacher use the storage's keyFunc, rather than diverging implementations more. Right now, the storage key calculation just happens to match between the cacher and the storage.